integration/kotlinx-coroutines-androidx-lifecycle/src/Lifecycle.kt

```diff
@@ -0,0 +1,92 @@
+
+import androidx.lifecycle.GenericLifecycleObserver
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.Lifecycle.State.DESTROYED
+import androidx.lifecycle.Lifecycle.State.INITIALIZED
+import androidx.lifecycle.LifecycleOwner
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import java.util.concurrent.ConcurrentHashMap
+
+/**
+ * Returns a [CoroutineScope] that uses [Dispatchers.Main] by default, and that is cancelled when
+ * the [Lifecycle] reaches [Lifecycle.State.DESTROYED] state.
+ *
+ * Note that this value is cached until the Lifecycle reaches the destroyed state.
+ */
+val Lifecycle.coroutineScope: CoroutineScope
+    get() = lifecycleCoroutineScopes[this] ?: job.let { job ->
+        val newScope = CoroutineScope(job + Dispatchers.Main)
+        if (job.isActive) {
+            lifecycleCoroutineScopes[this] = newScope
+            job.invokeOnCompletion { _ -> lifecycleCoroutineScopes -= this }
+        }
+        newScope
+    }
+
+/**
+ * Calls [Lifecycle.coroutineScope] for the [Lifecycle] of this [LifecycleOwner].
+ *
+ * This is an inline property, just there for convenient usage from any [LifecycleOwner],
+ * like FragmentActivity, AppCompatActivity, Fragment and LifecycleService.
+ */
+inline val LifecycleOwner.coroutineScope get() = lifecycle.coroutineScope
+
+/**
+ * Returns a [CoroutineScope] that uses [Dispatchers.Main] by default, and that will be cancelled as
+ * soon as this [Lifecycle] [currentState][Lifecycle.getCurrentState] is no longer
+ * [at least][Lifecycle.State.isAtLeast] the passed [activeWhile] state.
+ *
+ * **Beware**: if the current state is lower than the passed [activeWhile] state, you'll get an
+ * already cancelled scope.
+ */
+fun Lifecycle.createScope(activeWhile: Lifecycle.State): CoroutineScope {
+    if (activeWhile == DESTROYED) return coroutineScope
+    return CoroutineScope(createJob(activeWhile) + Dispatchers.Main)
+}
+
+/**
+ * Returns a [Job] that will be cancelled as soon as the [Lifecycle] reaches
+ * [Lifecycle.State.DESTROYED] state.
+ *
+ * Note that this value is cached until the Lifecycle reaches the destroyed state.
+ *
+ * You can use this job for custom [CoroutineScope]s, or as a parent [Job].
+ */
+val Lifecycle.job: Job
+    get() = lifecycleJobs[this] ?: createJob().also {
+        if (it.isActive) {
+            lifecycleJobs[this] = it
+            it.invokeOnCompletion { _ -> lifecycleJobs -= this }
+        }
+    }
+
+/**
+ * Creates a [Job] that will be cancelled as soon as this [Lifecycle]
+ * [currentState][Lifecycle.getCurrentState] is no longer [at least][Lifecycle.State.isAtLeast] the
+ * passed [activeWhile] state.
+ *
+ * **Beware**: if the current state is lower than the passed [activeWhile] state, you'll get an
+ * already cancelled job.
+ */
+fun Lifecycle.createJob(activeWhile: Lifecycle.State = INITIALIZED): Job {
+    require(activeWhile != Lifecycle.State.DESTROYED) {
```

Comment:
Maybe better would be return canceled Job? This would be consistent with coroutineScope behavior (just cancel job and all coroutines, do not throw an exception)

Reply:
Picture this: 'Give me a job that is active while the lifecycle is in destroyed state."

This doesn't make sense, or if you still want it, then, it's just an always active, which should not be tied to the `Lifecycle` in the first place.

Reply:
This is doesn't make sense if such call made in some synchronous code and obvious for a developer.
But in the asynchronous world is not so straightforward, it's usual problem to call this function is some async code and cancellation is not instant (even with coroutines, .cancel() is not suspending or blocking function), so it's easy to imagine cases when you will get this exception for edge cases and get rare IllegalStateExceptions instead of gracefully cancel coroutines where this job is used

Reply:
@gildor I don't understand why asynchronous code would make it so the function is called with `DESTROYED`. Here, this argument is just plain forbidden, and fails fast.

Reply:
Sorry, misunderstood original code, you right of course